posts/2021-03-13-microprofile-config-2.0.adoc

```diff
@@ -0,0 +1,331 @@
+config.getValue("empty.array.property", String[].class); // Will thorw a `NoSuchElementException`
+----
++
+However these values can be retrieved "optionally":
++
+[source,java]
+----
+@Inject
+@ConfigProperty(name = "empty.property")
+Optional<String> emptyProperty; // returns "Optional.empty"
+
+@Inject
+@ConfigProperty(name = "empty.array.property")
+Optional<String[]> emptyArrayProperty; // returns "Optional.empty"
+----
++
+and
++
+[source,java]
+----
+Config config = ConfigProvider.getConfig();
+config.getOptionalValue("empty.property", String.class); // returns “Optional.empty"
+config.getOptionalValue("empty.array.property", String[].class); // returns “Optional.empty"
+----
+
+
+* This means that link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/Config.html#getValue-java.lang.String-java.lang.Class-[Config.getValue()] will never return null; a `NoSuchElementException` will be thrown if the property:
+
+** Is not defined
+** Is defined as an empty String (`""`)
+** Is converted to `null` (considered to be "empty") by its Converter
+
+[#Expanding-Config-API]
+=== Expanding the Config API
+Two new methods have been added to the link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/Config.html[Config API class]:
+
+1. link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/Config.html#getValues-java.lang.String-java.lang.Class-[Config.getValues()]: `List<T> Config.getValues(String propertyName, Class<T> propertyType)`
+
+2. link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/Config.html#getOptionalValues-java.lang.String-java.lang.Class-[Config.getOptionalValues()]: `Optional <List<T>>Config.getOptionalValues(String propertyName, Class<T> propertyType)`
+
+The methods have been added to enable you to retrieve multi-valued config property values as lists instead of arrays. The methods return the resolved property values with the specified `propertyType` for the specified `propertyName`. For example:
+
+[source,java]
+----
+Config config = ConfigProvider.getConfig();
+config.getOptionalValues("ports", Integer.class) // returns "Optional[[9080, 9081, 9082]]"
+----
+
+[#More-Optional-Converter]
+=== More Optional Converters
+`OptinalInt`, `OptionalLong` and `OptionalDouble` are now provided as link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/spi/Converter.html#built_in_converters[built-in Converters]. The new Converters can be used like any of the other built-in Converters; converting injected config property values to a defined type:
+
+[source,java]
+----
+@Inject
+@ConfigProperty(name = "server.port")
+OptionalInt optionalServerPort; // returns “9080"
+----
+
+[#Incompatibility-changes]
+== Heads up! Incompatibility Changes
```

Comment:
"Incompatibility Changes" doesn't sound right to me ... "Incompatible Changes" is better?

Reply:
I'm not sure. I think I prefer "Incompatibility Changes". Would be happy to hear what anyone else thinks